api/tests/src/services/applications/test_submit_application.py

```diff
@@ -1,62 +1,144 @@
+        grace_period=0,
+    )
+    application1 = ApplicationFactory.create(competition=competition1)
+    db_session.commit()
+
+    # Should not raise any exception
+    validate_competition_open(application1)
+
+    # Past closing date but within grace period
+    competition2 = CompetitionFactory.create(
+        closing_date=current_date - timedelta(days=2),
+        grace_period=3,
+    )
+    application2 = ApplicationFactory.create(competition=competition2)
+    db_session.commit()
+
+    # Should not raise any exception
+    validate_competition_open(application2)
+
+
+@freeze_time(TEST_DATE)
+def test_validate_competition_open_failure(enable_factory_create, db_session):
+    """Test that validating a closed competition raises an error."""
+    current_date = date.fromisoformat(TEST_DATE)
+
+    # Create a competition with a past closing date (5 days ago) and a grace period of 3 days
+    # This means the competition has been closed for 2 days
+    competition = CompetitionFactory.create(
+        closing_date=current_date - timedelta(days=5),
+        grace_period=3,
+    )
+    application = ApplicationFactory.create(competition=competition)
+    db_session.commit()
+
+    with pytest.raises(apiflask.exceptions.HTTPError) as excinfo:
+        validate_competition_open(application)
+
+    assert excinfo.value.status_code == 422
+    assert "Cannot submit application - competition is closed" in excinfo.value.message
+    assert (
+        excinfo.value.extra_data["validation_issues"][0].type
+        == ValidationErrorType.COMPETITION_ALREADY_CLOSED
+    )
+
+
+@freeze_time(TEST_DATE)
+def test_validate_competition_with_no_closing_date(enable_factory_create, db_session):
+    """Test that validating a competition with no closing date succeeds."""
+    competition = CompetitionFactory.create(closing_date=None)
+    application = ApplicationFactory.create(competition=competition)
+    db_session.commit()
+
+    # Should not raise any exception
+    validate_competition_open(application)
+
 
-    # Verify status hasn't changed
-    db_session.refresh(application)
-    assert application.application_status == initial_status
+# Tests for the main submit_application function
+def test_submit_application_success(enable_factory_create, db_session):
+    """Test successful submission of an application in IN_PROGRESS state."""
+    today = date.today()
+    competition = CompetitionFactory.create(
+        closing_date=today + timedelta(days=1),
+        grace_period=3,
+    )
+    application = ApplicationFactory.create(
+        application_status=ApplicationStatus.IN_PROGRESS, competition=competition
+    )
+    db_session.commit()  # Commit initial state
+
+    updated_application = submit_application(db_session, application.application_id)
+    db_session.commit()  # Commit the change made by the service
+
+    db_session.refresh(updated_application)  # Refresh to get the latest state from DB
```

Comment:
Factories already commit as part of create. Any case of `SomeFactory.create(...)` followed by `db_session.commit()` isn't necessary.

What is refreshing from the DB doing here when the function returns the DB object?

```suggestion
    application = ApplicationFactory.create(
        application_status=ApplicationStatus.IN_PROGRESS, competition=competition
    )

    updated_application = submit_application(db_session, application.application_id)
```

Reply:
Yeah, I thought those references were removed. I was debugging another issue in the code that may have been related to a record not saving in the DB, and added a commit to debug that wasn't properly removed